tests/separator.py

```diff
@@ -0,0 +1,502 @@
+
+        # THEN
+        self.ca.assert_that_pv_is("DAQ:VOLT:SP", 20. * DAQ_VOLT_WRITE_SCALE_FACTOR)
+
+    @parameterized.expand(parameterized_list(voltage_values))
+    def test_that_WHEN_set_THEN_the_voltage_changes(self, _, value):
+        # WHEN
+        self.ca.set_pv_value("VOLT:SP", value)
+
+        # THEN
+        self.ca.assert_that_pv_is_number("VOLT", value, MARGIN_OF_ERROR)
+
+    @parameterized.expand(parameterized_list(voltage_values_which_give_alarms))
+    def test_that_WHEN_voltage_out_of_range_THEN_alarm_raised(self, _, value):
+        # WHEN
+
+        self.ca.set_pv_value("VOLT:SP", value)
+
+        # THEN
+        self.ca.assert_that_pv_alarm_is("VOLT", ChannelAccess.Alarms.MAJOR)
+
+    def test_that_GIVEN_voltage_in_range_WHEN_setpoint_is_above_range_THEN_setpoint_is_set_to_max_value(self):
+        # GIVEN
+        self.ca.set_pv_value("VOLT:SP", 30)
+        self.ca.assert_that_pv_is("VOLT", 30)
+
+        # WHEN
+        self.ca.set_pv_value("VOLT:SP", 215.)
+
+        # THEN
+        self.ca.assert_that_pv_is("VOLT:SP", MAX_SEPARATOR_VOLT)
+        self.ca.assert_that_pv_is("VOLT", MAX_SEPARATOR_VOLT)
+
+    def test_that_GIVEN_voltage_in_range_WHEN_setpoint_is_below_range_THEN_setpoint_is_set_to_min_value(self):
+        # GIVEN
+        self.ca.set_pv_value("VOLT:SP", 30)
+        self.ca.assert_that_pv_is("VOLT", 30)
+
+        # WHEN
+        self.ca.set_pv_value("VOLT:SP", -50)
+
+        # THEN
+        self.ca.assert_that_pv_is("VOLT", MIN_SEPARATOR_VOLT)
+        self.ca.assert_that_pv_is("VOLT:SP", MIN_SEPARATOR_VOLT)
+
+
+class CurrentTests(unittest.TestCase):
+    # These current testing values are uncalibrated values from the DAQ lying between 0 and 10.
+    current_values = [0, 1.33333, 5e1, 10e-3, 10]
+    current_values_which_give_alarms = [-2, 0, 10, 11]
+
+    def setUp(self):
+        self.ca = ChannelAccess(20, device_prefix=DEVICE_PREFIX)
+        shared_setup(self.ca)
+        self._simulate_current(0)
+
+    def _simulate_current(self, current):
+        curr_array = [current] * 1000
+        self.ca.set_pv_value("DAQ:CURR:WV:SIM", curr_array)
+
+    @parameterized.expand(parameterized_list(current_values))
+    def test_that_GIVEN_current_value_THEN_calibrated_current_readback_changes(self, _, value):
+        # GIVEN
+        self._simulate_current(value)
+
+        self.ca.assert_that_pv_is_number("CURR", value * DAQ_CURR_READ_SCALE_FACTOR, MARGIN_OF_ERROR)
+
+    @parameterized.expand(parameterized_list(current_values_which_give_alarms))
+    def test_that_WHEN_current_is_out_of_range_THEN_alarm_raised(self, _, value):
+        # WHEN
+        self._simulate_current(value)
+
+        # THEN
+        self.ca.assert_that_pv_alarm_is("CURR", ChannelAccess.Alarms.MAJOR)
+
+
+class StabilityTests(unittest.TestCase):
+    STOP_DATA_THREAD = threading.Event()
+
+    def setUp(self):
+        self.ca = ChannelAccess(20, device_prefix=DEVICE_PREFIX)
+        shared_setup(self.ca)
```

Comment:
Because two of the classes use (subtly) different setUp functions, I didn't use the `add_method` decorator as I didn't want to have `setUp(self)` defined differently in main and in these 'different' classes